lib/src/test/java/io/ably/lib/transport/HostsTest.java

```diff
@@ -0,0 +1,271 @@
+package io.ably.lib.transport;
+
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.isEmptyOrNullString;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+import io.ably.lib.types.AblyException;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+import java.util.List;
+import org.junit.Test;
+
+import io.ably.lib.types.ClientOptions;
+
+public class HostsTest {
+
+    /**
+     * Tests for Hosts class (fallback hosts).
+     */
+    @Test
+    public void hosts_fallback() throws AblyException {
+        ClientOptions options = new ClientOptions();
+        Hosts hosts = new Hosts(Defaults.HOST_REALTIME, Defaults.HOST_REALTIME, options);
+        String host = hosts.getFallback(Defaults.HOST_REALTIME);
+        /* Expect given fallback host string is (relatively) valid */
+        assertThat(host, not(isEmptyOrNullString()));
+        /* Expect multiple calls will provide different (relatively) valid fallback hosts */
+        String host2 = hosts.getFallback(host);
+        assertThat(host2, not(isEmptyOrNullString()));
+        assertThat(host2, not(host));
+        /* Expect a null, when we requested more than available fallback hosts */
+        for (int i = Defaults.HOST_FALLBACKS.length - 1; i > 0; i--) {
```

Comment:
I would generally avoid int iterations as much as possible. 
Can you check if the java version supports?

```for (String host : Defaults.HOST_FALLBACKS) { }```
or you can use collections API for iterations

Reply:
Also, I am not sure, what this for loop exactly does 🤔

Reply:
There are was a bunch of tests which were there before and it seems that I wrote them, but I've just moved this file. Yes, the logic is more complicated than it should be, but I checked the tests and they look good to me. I would be glad to make these tests nicer but in a separate pull request.

Reply:
Okay, makes sense now. But, I am afraid if we will have the capacity to raise another PR (just to fix tests). I think it's better if we fix it as a part of this PR ( Till the soup is HOTT 😋). Also, I strongly feel the test file is super relevant with modifications made to the code, so makes sense to modify existing tests by any means necessary 😃

Reply:
Also, make sure you add missing tests with respect to Js PR.